### ISSUE TITLE
fundering deeplearning

### ISSUE THREAD

**Issue by groenklimaat** [groenklimaat/Automatically-archive-by-AI#10]
#include <math.h>
/******************************************************************
* Network Configuration - customized per network
******************************************************************/
// Node parameters
const int patternCount = 4; // Number of input patterns
const int inputNodes = 2; // Number of input nodes
const int hiddenNodes = 2; / Number of hidden nodes
const int outputNodes = 1; // Number of output nodes
// Study parameters
const float learningRate = 0.3; // Adjusts how much of the error is actually backpropagated.
const float momentum = 0.1; // Adjusts how much the results of the previous iteration affect the current iteration.
const float initialWeightMax = 0.4; // Sets the maximum starting value for weights.
const float success = 0.01; // The threshold for error at which the training will be considered successful.
const int REPORTITERATIONS = 50; // Report to serial monitor every n REPORTITERATIONS
// inputData Pattern/Training data
const byte inputData[patternCount][inputNodes] = {
{ 0, 0},
{ 1, 0},
{ 0, 1},
{ 1, 1}
};
// Wanted output Pattern
const byte targetData[patternCount][outputNodes] = {0, 1, 1, 0 };
/******************************************************************
* End Network Configuration
******************************************************************/
int i, j, p, q, r; // Counters
int reportAfterIterations; // Counter for reporting after n iterations
int randomizedIndex[patternCount]; // To shuffle and randomize training pattern
long trainingCycle; // Keeps track of number of training cycles
float randomized; // To randomize weights
float error; // For the error-calculation of the output
float accum; // For accumulated weights
// Arrays for nodes and weights between nodes
float hidden[hiddenNodes];
float output[outputNodes];
float hiddenWeights[inputNodes+1][hiddenNodes];
float outputWeights[hiddenNodes+1][outputNodes];
float hiddenDelta[hiddenNodes];
float outputDelta[outputNodes];
float changeHiddenWeights[inputNodes+1][hiddenNodes];
float changeOutputWeights[hiddenNodes+1][outputNodes];
void setup(){
Serial.begin(115200);
randomSeed(analogRead(3)); // Random input from A3 to startup random number generator
// Necessary for later use of random()
reportAfterIterations = 1; // Initialize report counter
for( p = 0 ; p < patternCount ; p++ ) { // Initialize array
randomizedIndex[p] = p ;}
}
void loop (){
/******************************************************************
* Initialize hiddenWeights and changeHiddenWeights
******************************************************************/
for( i = 0 ; i < hiddenNodes ; i++ ) {
for( j = 0 ; j <= inputNodes ; j++ ) { // '<=' to implement one extra bias node
changeHiddenWeights[j][i] = 0.0 ; // Initialize hidden weights
randomized = float(random(100))/100; // Random number between 0..100/100 = float >=0 && <=1
hiddenWeights[j][i] = 2.0 * ( randomized - 0.5 ) * initialWeightMax ; // Randomize weights before training
}
}
/******************************************************************
* Initialize outputWeights and changeOutputWeights
******************************************************************/
for( i = 0 ; i < outputNodes ; i ++ ) {
for( j = 0 ; j <= hiddenNodes ; j++ ) { // '<=' to implement one extra bias node
changeOutputWeights[j][i] = 0.0 ; // Initialize hidden weights
randomized = float(random(100))/100; // Random number between 0..100/100 = float >=0 && <=1
outputWeights[j][i] = 2.0 * ( randomized - 0.5 ) * initialWeightMax ; // Randomize weights before training
}
}
Serial.println("Initial/Untrained outputs: ");
toTerminal();
/******************************************************************
* Begin training
******************************************************************/
// Max trainingcycles = unsigned long = 31 bits
for( trainingCycle = 1 ; trainingCycle < 2147483647 ; trainingCycle++) {
/* Every training cycle contains the following blocks :
- Randomize order of training patterns
- Compute hidden layer activation functions (sigmoid function)
- Compute outut layers activation and error
- Backpropagate errors to hidden layer
- Update weights from input nodes to hidden node(s)
- Update weights from hidden nodes to output node(s) */
/******************************************************************
* randomizedmize order of training patterns
******************************************************************/
// For all patterns, switch randomizedIndex[p] with randomizedIndex[random q]
for( p = 0 ; p < patternCount ; p++) {
q = random(patternCount);
// q is a random number between 0 en number of patterns
r = randomizedIndex[p] ;
// r becomes indexvalue of index p
randomizedIndex[p] = randomizedIndex[q] ; // index p gets indexvalue of index q
randomizedIndex[q] = r ;
// inde q gets old indexvalue of p (equals r)
}
error = 0.0 ;
/******************************************************************
* Cycle through each training pattern in the randomized order
******************************************************************/
for( q = 0 ; q < patternCount ; q++ ) {p = randomizedIndex[q];
/******************************************************************
* Compute hidden layer activations
******************************************************************/
for( i = 0 ; i < hiddenNodes ; i++ ) { // For every hidden node, accumulate the weights of all connected input nodes
accum = hiddenWeights[inputNodes][i] ;
for( j = 0 ; j < inputNodes ; j++ ) {
accum += inputData[p][j] * hiddenWeights[j][i] ;
}
hidden[i] = 1.0/(1.0 + exp(-accum)) ; // Sigmoid function to transform input to a number between 0 and 1
}
/******************************************************************
* Compute output layer activations and calculate errors
******************************************************************/
// For every output node, accumulate the weights of all connected hidden nodes
for( i = 0 ; i < outputNodes ; i++ ) {
accum = outputWeights[hiddenNodes][i] ;
for( j = 0 ; j < hiddenNodes ; j++ ) {
accum += hidden[j] * outputWeights[j][i] ;
}
output[i] = 1.0/(1.0 + exp(-accum)) ; // Sigmoïde functie to transform input to a number between 0 and 1
// Calculate difference between wanted target data and calculated output data
outputDelta[i] = (targetData[p][i] - output[i]) * output[i] * (1.0 - output[i]) ;
// Calculate error coefficient
error += 0.5 * (targetData[p][i] - output[i]) * (targetData[p][i] - output[i]) ;
}
/******************************************************************
* Backpropagate errors to hidden layer
******************************************************************/
for( i = 0 ; i < hiddenNodes ; i++ ) {
accum = 0.0 ;
for( j = 0 ; j < outputNodes ; j++ ) {
accum += outputWeights[i][j] * outputDelta[j] ;
}
hiddenDelta[i] = accum * hidden[i] * (1.0 - hidden[i]) ;
}
/******************************************************************
* Update Inner-->hidden Weights
******************************************************************/
for( i = 0 ; i < hiddenNodes ; i++ ) {
changeHiddenWeights[inputNodes][i] = learningRate * hiddenDelta[i] + momentum *
changeHiddenWeights[inputNodes][i] ;
hiddenWeights[inputNodes][i] += changeHiddenWeights[inputNodes][i] ;
for( j = 0 ; j < inputNodes ; j++ ) {
changeHiddenWeights[j][i] = learningRate * inputData[p][j] * hiddenDelta[i] + momentum *
changeHiddenWeights[j][i];
hiddenWeights[j][i] += changeHiddenWeights[j][i] ;
}
}
/******************************************************************
* Update hidden-->output Weights
******************************************************************/for( i = 0 ; i < outputNodes ; i ++ ) {
changeOutputWeights[hiddenNodes][i] = learningRate * outputDelta[i] + momentum *
changeOutputWeights[hiddenNodes][i] ;
outputWeights[hiddenNodes][i] += changeOutputWeights[hiddenNodes][i] ;
for( j = 0 ; j < hiddenNodes ; j++ ) {
changeOutputWeights[j][i] = learningRate * hidden[j] * outputDelta[i] + momentum * changeOutputWeights[j][i] ;
outputWeights[j][i] += changeOutputWeights[j][i] ;
}
}
}
/******************************************************************
* Every REPORTITERATIONS cycles send data to terminal for display
******************************************************************/
reportAfterIterations = reportAfterIterations - 1;
if (reportAfterIterations == 0)
{
Serial.println();
Serial.println();
Serial.print ("trainingCycle: ");
Serial.print (trainingCycle);
Serial.print (" error = ");
Serial.println (error, 5);
toTerminal();
if (trainingCycle==1)
{
reportAfterIterations = REPORTITERATIONS-1;
}
else
{
reportAfterIterations = REPORTITERATIONS;
}
}
/******************************************************************
* If error rate is less than pre-determined threshold then end
******************************************************************/
if( error < success ) break ;
}
Serial.println ();
Serial.println();
Serial.print ("trainingCycle: ");
Serial.print (trainingCycle);
Serial.print (" error = ");
Serial.println (error, 5);
toTerminal();
Serial.println ();
Serial.println ();
Serial.println ("Training Set Solved! ");
Serial.println ("--------");
Serial.println ();
Serial.println ();
while(1) {};reportAfterIterations = 1;
}
void toTerminal()
{
for( p = 0 ; p < patternCount ; p++ ) {
Serial.println();
Serial.print (" Training Pattern: ");
Serial.println (p);
Serial.print (" inputData ");
for( i = 0 ; i < inputNodes ; i++ ) {
Serial.print (inputData[p][i], DEC);
Serial.print (" ");
}
Serial.print (" targetData ");
for( i = 0 ; i < outputNodes ; i++ ) {
Serial.print (targetData[p][i], DEC);
Serial.print (" ");
}
/******************************************************************
* Compute hidden layer activations
******************************************************************/
for( i = 0 ; i < hiddenNodes ; i++ ) {
accum = hiddenWeights[inputNodes][i] ;
for( j = 0 ; j < inputNodes ; j++ ) {
accum += inputData[p][j] * hiddenWeights[j][i] ;
}
hidden[i] = 1.0/(1.0 + exp(-accum)) ;
}
/******************************************************************
* Compute output layer activations and calculate errors
******************************************************************/
for( i = 0 ; i < outputNodes ; i++ ) {
accum = outputWeights[hiddenNodes][i] ;
for( j = 0 ; j < hiddenNodes ; j++ ) {
accum += hidden[j] * outputWeights[j][i] ;
}
output[i] = 1.0/(1.0 + exp(-accum)) ;
}
Serial.print (" output ");
for( i = 0 ; i < outputNodes ; i++ ) {
Serial.print (output[i], 5);
Serial.print (" ");
}
}
}
